### PR TITLE
Putting it all together: the information flow analysis algorithm

### PR DESCRIPTION
**Pull request by bgogul**
This PR implements the information flow analysis algorithm. This PR establishes parity with the type script implementation for claims and checks related to semantic labels on handles. 

I have only added integration tests based on the tests from typescript land. It will be useful to have unit tests for the individual transfer functions, which I will do in subsequent PRs.

### CODE REVIEW

File: java/arcs/core/analysis/InformationFlow.kt [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+import arcs.core.data.InformationFlowLabel.SemanticTag
+import arcs.core.data.Recipe
+import arcs.core.data.Recipe.Particle
+import arcs.core.util.TaggedLog
+import java.util.BitSet
+
+/**
+ * Information flow analysis of a [Recipe] using the labels in the particle checks and claims.
+ *
+ * For the time being, we pass in the ingress handles.
```
Comment (csilvestrini) [inline java/arcs/core/analysis/InformationFlow.kt:30]: Might be worth adding a TODO comment here

File: java/arcs/core/analysis/InformationFlow.kt [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+
+        // For read-write connections, apply any claims that are present. This is necessary as
+        // the claim on a read-write connection immediately manifests on the read connection.
+        particleClaims[particle]?.let { claims ->
+            initialValues.applyClaims(
+                claims.filterIsInstance<Claim.Assume>()
+                    .filter {
+                        val accessPathRoot =
+                            it.accessPath.root as? AccessPath.Root.HandleConnection
+                        accessPathRoot?.connectionSpec?.isReadWrite() == true
```
Comment (csilvestrini) [inline java/arcs/core/analysis/InformationFlow.kt:135]: Nit: you don't need the `== true`

File: javatests/arcs/core/analysis/InformationFlowTest.kt [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        val violation: List<String> = emptyList()
+    )
+
+    /** Returns null if the recipe is valid. Otherwise, returns the first violating check. */
+    private fun getViolatingChecks(test: String, ingresses: List<String>): List<String> {
+        val recipe = parseManifestWithSingleRecipe(test)
+        val result = InformationFlow.computeLabels(recipe, ingresses)
+        return result.checks.flatMap { (particle, checks) ->
+            checks.filter {
+                check -> !result.verify(particle, check)
```
Comment (csilvestrini) [inline javatests/arcs/core/analysis/InformationFlowTest.kt:56]: Nit: `check ->` goes on the previous line

File: javatests/arcs/core/analysis/InformationFlowTest.kt [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            }.map { it.asString() }
+        }
+    }
+
+    @Test
+    fun successDFA() {
+        val tests = listOf(
+            DFATest("ok-directly-satisfied", listOf("P1")),
```
Comment (csilvestrini) [inline javatests/arcs/core/analysis/InformationFlowTest.kt:64]: Optional: you could put a magic comment in the file that indicates the ingress points, e.g. something like this:

```
//#ingress = P1
```

It wouldn't be hard to parse, and would let you inline this directly in the test file.

File: javatests/arcs/core/analysis/InformationFlowTest.kt [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        tests.forEach { (test, ingresses, _) ->
+            assertWithMessage("Test '$test' should be valid!")
+                .that(getViolatingChecks(test, ingresses)).isEmpty()
+        }
+    }
+
+    @Test
+    fun failDFA() {
+        val tests = listOf(
+            DFATest("fail-different-tag", listOf("P1"), listOf("hc:P2.bar is trusted")),
```
Comment (csilvestrini) [inline javatests/arcs/core/analysis/InformationFlowTest.kt:85]: Similarly here:

```
//#failure = hc:P2.bar is trusted
```

File: java/arcs/core/analysis/InformationFlow.kt (hunk 2 above)
Comment (bgogul) [reply java/arcs/core/analysis/InformationFlow.kt:135]: `== true` is needed as the expression is a nullable Boolean. See https://kotlinlang.org/docs/reference/idioms.html#consuming-a-nullable-boolean

File: javatests/arcs/core/analysis/InformationFlowTest.kt (hunk 4 above)
Comment (bgogul) [reply javatests/arcs/core/analysis/InformationFlowTest.kt:64]: Good idea. Let do this in a following PR? Doing this here would pull in all the test files as well.

File: javatests/arcs/core/analysis/InformationFlowTest.kt (hunk 5 above)
Comment (bgogul) [reply javatests/arcs/core/analysis/InformationFlowTest.kt:85]: +1 WIll do in a separate PR.

File: java/arcs/core/analysis/InformationFlow.kt (hunk 1 above)
Comment (bgogul) [reply java/arcs/core/analysis/InformationFlow.kt]: Done.
